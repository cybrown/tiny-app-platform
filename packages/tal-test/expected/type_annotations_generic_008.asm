main()
  entry:
    FunctionRef       name: run_0
    DeclareLocal      name: run, mutable: false, hasInitialValue: true
    Pop               inBlock: false
    Literal           "8"
    FunctionRef       name: func_1
    Literal           2
    MakeArray
    Literal           0
    MakeRecord
    Local             run
    Call
    Pop               inBlock: false
    Literal           8
    FunctionRef       name: func_2
    Literal           2
    MakeArray
    Literal           0
    MakeRecord
    Local             run
    Call
run_0(n, f)
  entry:
    Literal           0
    MakeArray
    Literal           "value"
    Local             n
    Literal           1
    MakeRecord
    Local             f
    Call
    MakeArrayForBlock count: 1
func_1(value)
  entry:
    Local             value
    Literal           "a"
    Intrinsic         INTRINSIC_ADD
    Literal           "toto"
    Intrinsic         INTRINSIC_EQUAL
func_2(value)
  entry:
    Local             value
    Literal           "a"
    Intrinsic         INTRINSIC_ADD
    Literal           "toto"
    Intrinsic         INTRINSIC_EQUAL
