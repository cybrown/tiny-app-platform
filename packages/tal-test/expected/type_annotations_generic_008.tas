fun run(n: T, f: (value: T) => U): U { f(value: n) }

run("8", value => value + "a" == "toto")
run(8, value => value + "a" == "toto")
