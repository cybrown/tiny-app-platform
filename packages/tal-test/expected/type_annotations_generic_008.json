[
  {
    "kind": "DeclareLocal",
    "mutable": false,
    "name": "run",
    "value": {
      "kind": "Function",
      "body": {
        "kind": "Block",
        "children": [
          {
            "kind": "Call",
            "args": [
              {
                "kind": "NamedArgument",
                "name": "value",
                "value": {
                  "kind": "Local",
                  "name": "n"
                }
              }
            ],
            "value": {
              "kind": "Local",
              "name": "f"
            }
          }
        ]
      },
      "parameters": [
        {
          "name": "n",
          "type": {
            "kind": "named",
            "name": "T"
          }
        },
        {
          "name": "f",
          "type": {
            "kind": "function",
            "parameters": [
              {
                "name": "value",
                "type": {
                  "kind": "named",
                  "name": "T"
                }
              }
            ],
            "returnType": {
              "kind": "named",
              "name": "U"
            }
          }
        }
      ],
      "returnType": {
        "kind": "named",
        "name": "U"
      },
      "genericParameters": [
        {
          "name": "T"
        },
        {
          "name": "U"
        }
      ]
    }
  },
  {
    "kind": "Call",
    "args": [
      {
        "kind": "PositionalArgument",
        "value": "8"
      },
      {
        "kind": "PositionalArgument",
        "value": {
          "kind": "Function",
          "body": {
            "kind": "BinaryOperator",
            "operator": "==",
            "right": "toto",
            "left": {
              "kind": "BinaryOperator",
              "operator": "+",
              "right": "a",
              "left": {
                "kind": "Local",
                "name": "value"
              }
            }
          },
          "parameters": [
            {
              "name": "value"
            }
          ]
        }
      }
    ],
    "value": {
      "kind": "Local",
      "name": "run"
    }
  },
  {
    "kind": "Call",
    "args": [
      {
        "kind": "PositionalArgument",
        "value": 8
      },
      {
        "kind": "PositionalArgument",
        "value": {
          "kind": "Function",
          "body": {
            "kind": "BinaryOperator",
            "operator": "==",
            "right": "toto",
            "left": {
              "kind": "BinaryOperator",
              "operator": "+",
              "right": "a",
              "left": {
                "kind": "Local",
                "name": "value"
              }
            }
          },
          "parameters": [
            {
              "name": "value"
            }
          ]
        }
      }
    ],
    "value": {
      "kind": "Local",
      "name": "run"
    }
  }
]